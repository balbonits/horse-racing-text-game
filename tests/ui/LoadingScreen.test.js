/**
 * Loading Screen System Tests
 * 
 * Tests the loading screen animations, messages, and state transitions
 * with snapshot testing for visual consistency.
 */

const LoadingScreen = require('../../src/ui/screens/LoadingScreen');
const ColorThemeManager = require('../../src/ui/ColorThemeManager');

describe('Loading Screen System', () => {
  let loadingScreen;
  let colorManager;
  let consoleOutput;
  let originalConsoleLog;
  let originalConsoleClear;

  beforeEach(() => {
    // Setup color manager
    colorManager = new ColorThemeManager();
    loadingScreen = new LoadingScreen(colorManager);

    // Mock console output for testing
    consoleOutput = [];
    originalConsoleLog = console.log;
    originalConsoleClear = console.clear;

    console.log = jest.fn((...args) => {
      consoleOutput.push(args.join(' '));
    });
    console.clear = jest.fn();
  });

  afterEach(() => {
    // Restore console
    console.log = originalConsoleLog;
    console.clear = originalConsoleClear;
  });

  describe('Animation Selection', () => {
    test('should select track animation for race transitions', () => {
      expect(loadingScreen.selectAnimation('training', 'race_preview')).toBe('track');
      expect(loadingScreen.selectAnimation('race_running', 'race_results')).toBe('track');
      expect(loadingScreen.selectAnimation('race_results', 'podium')).toBe('track'); // Changed test case
    });

    test('should select hoofprints animation for training transitions', () => {
      expect(loadingScreen.selectAnimation('character_creation', 'training')).toBe('progress'); // Correct expectation - character_creation takes priority
      expect(loadingScreen.selectAnimation('race_results', 'training')).toBe('track'); // race_results contains 'race', so track takes priority
      expect(loadingScreen.selectAnimation('help', 'training')).toBe('hoofprints'); // Pure training transition
    });

    test('should select horse animation for tutorial transitions', () => {
      expect(loadingScreen.selectAnimation('main_menu', 'tutorial')).toBe('horse');
      expect(loadingScreen.selectAnimation('tutorial', 'tutorial_training')).toBe('horse');
      expect(loadingScreen.selectAnimation('tutorial_training', 'tutorial_race')).toBe('horse');
    });

    test('should select progress animation for character creation', () => {
      expect(loadingScreen.selectAnimation('main_menu', 'character_creation')).toBe('progress');
    });

    test('should default to spinner for other transitions', () => {
      expect(loadingScreen.selectAnimation('main_menu', 'help')).toBe('spinner');
      expect(loadingScreen.selectAnimation('help', 'main_menu')).toBe('spinner');
    });
  });

  describe('Transition Messages', () => {
    test('should have appropriate messages for main menu transitions', () => {
      expect(loadingScreen.transitionMessages['main_menu→character_creation']).toEqual([
        'Preparing the stables...',
        'Setting up your office...',
        'Meeting the staff...'
      ]);

      expect(loadingScreen.transitionMessages['main_menu→tutorial']).toEqual([
        'Loading tutorial...',
        'Preparing guided experience...',
        'Alex Morgan is on the way...'
      ]);
    });

    test('should have race-specific transition messages', () => {
      expect(loadingScreen.transitionMessages['race_preview→horse_lineup']).toEqual([
        'Entering the paddock...',
        'Saddling up...',
        'Meeting the other competitors...'
      ]);

      expect(loadingScreen.transitionMessages['strategy_select→race_running']).toEqual([
        'Taking positions...',
        'Gates loading...',
        'The crowd goes silent...'
      ]);
    });

    test('should have tutorial transition messages', () => {
      expect(loadingScreen.transitionMessages['tutorial→tutorial_training']).toEqual([
        'Starting tutorial...',
        'Alex Morgan arrives...',
        'Preparing lesson plan...'
      ]);
    });

    test('should fall back to generic messages for unknown transitions', () => {
      expect(loadingScreen.transitionMessages['any→main_menu']).toEqual([
        'Returning to main menu...',
        'Saving progress...',
        'See you at the stables...'
      ]);
    });
  });

  describe('Loading Animations', () => {
    test('should have multiple animation frame types', () => {
      expect(loadingScreen.loadingAnimations).toHaveProperty('dots');
      expect(loadingScreen.loadingAnimations).toHaveProperty('spinner');
      expect(loadingScreen.loadingAnimations).toHaveProperty('horse');
      expect(loadingScreen.loadingAnimations).toHaveProperty('track');
      expect(loadingScreen.loadingAnimations).toHaveProperty('hoofprints');
      expect(loadingScreen.loadingAnimations).toHaveProperty('progress');
    });

    test('should have proper frame sequences', () => {
      expect(loadingScreen.loadingAnimations.dots).toEqual(['   ', '.  ', '.. ', '...']);
      expect(loadingScreen.loadingAnimations.horse).toEqual(['🐎   ', ' 🐎  ', '  🐎 ', '   🐎', '  🐎 ', ' 🐎  ']);
      expect(loadingScreen.loadingAnimations.spinner).toHaveLength(10);
      expect(loadingScreen.loadingAnimations.track).toHaveLength(5);
    });
  });

  describe('Progress Bar Rendering', () => {
    test('should render progress bar with correct width and percentage', () => {
      const result = loadingScreen.renderProgressBar(0.5);
      expect(result).toMatch(/\[█{20}░{20}\]/); // 50% should be 20 filled, 20 empty
      expect(result).toContain('50%');
    });

    test('should render full progress bar', () => {
      const result = loadingScreen.renderProgressBar(1.0);
      expect(result).toMatch(/\[█{40}\]/); // 100% should be 40 filled
      expect(result).toContain('100%');
    });

    test('should render empty progress bar', () => {
      const result = loadingScreen.renderProgressBar(0);
      expect(result).toMatch(/\[░{40}\]/); // 0% should be 40 empty
      expect(result).toContain('0%');
    });
  });

  describe('Text Centering', () => {
    test('should center text properly', () => {
      // Mock terminal width
      const originalColumns = process.stdout.columns;
      process.stdout.columns = 80;

      const result = loadingScreen.centerText('Test');
      expect(result).toBe(' '.repeat(38) + 'Test'); // (80-4)/2 = 38 spaces

      process.stdout.columns = originalColumns;
    });

    test('should handle ANSI color codes in text centering', () => {
      const originalColumns = process.stdout.columns;
      process.stdout.columns = 80;

      const coloredText = '\x1b[32mTest\x1b[0m'; // Green "Test"
      const result = loadingScreen.centerText(coloredText);
      // Should calculate padding based on plain text length (4), not ANSI string length (10)
      expect(result).toBe(' '.repeat(38) + coloredText);

      process.stdout.columns = originalColumns;
    });
  });

  describe('Loading Screen Snapshots', () => {
    test('should render loading screen consistently', () => {
      loadingScreen.renderLoadingScreen('Loading test...', '🐎', 0.5);
      
      expect(consoleOutput.join('\n')).toMatchSnapshot('loading-screen-basic');
    });

    test('should render different animation frames consistently', () => {
      const frames = loadingScreen.loadingAnimations.track;
      const outputs = [];

      frames.forEach((frame, index) => {
        consoleOutput = []; // Clear output
        loadingScreen.renderLoadingScreen(`Track animation ${index}`, frame, index / frames.length);
        outputs.push(consoleOutput.join('\n'));
      });

      expect(outputs).toMatchSnapshot('track-animation-frames');
    });

    test('should render tutorial loading messages consistently', () => {
      const messages = loadingScreen.transitionMessages['tutorial→tutorial_training'];
      const outputs = [];

      messages.forEach((message, index) => {
        consoleOutput = []; // Clear output
        loadingScreen.renderLoadingScreen(message, '🐎', (index + 1) / messages.length);
        outputs.push(consoleOutput.join('\n'));
      });

      expect(outputs).toMatchSnapshot('tutorial-loading-messages');
    });

    test('should render race countdown consistently', () => {
      // Mock the displayRaceCountdown to capture frames
      const countdownFrames = [];
      const originalSetTimeout = setTimeout;
      
      // Mock setTimeout to capture frames synchronously
      global.setTimeout = (callback) => {
        callback();
        return {};
      };

      loadingScreen.displayRaceCountdown().then(() => {
        // Restore original setTimeout
        global.setTimeout = originalSetTimeout;
      });

      expect(consoleOutput).toMatchSnapshot('race-countdown-sequence');
    });
  });

  describe('Quick Display Methods', () => {
    test('should display quick message', async () => {
      const promise = loadingScreen.displayQuickMessage('Quick test', 100);
      
      expect(console.clear).toHaveBeenCalled();
      expect(consoleOutput.join('\n')).toContain('Quick test');
      
      await promise;
    });

    test('should display error message', async () => {
      const promise = loadingScreen.displayError('Test error', 100);
      
      expect(console.clear).toHaveBeenCalled();
      expect(consoleOutput.join('\n')).toContain('❌ Error ❌');
      expect(consoleOutput.join('\n')).toContain('Test error');
      
      await promise;
    });

    test('should display success message', async () => {
      const promise = loadingScreen.displaySuccess('Test success', 100);
      
      expect(console.clear).toHaveBeenCalled();
      expect(consoleOutput.join('\n')).toContain('✅ Success! ✅');
      expect(consoleOutput.join('\n')).toContain('Test success');
      
      await promise;
    });
  });

  describe('Integration with ColorThemeManager', () => {
    test('should apply colors when color manager is provided', () => {
      const coloredLoadingScreen = new LoadingScreen(colorManager);
      coloredLoadingScreen.renderLoadingScreen('Colored test', '🐎', 0.5);
      
      // Check if color methods were called or if specific coloring occurred
      const output = consoleOutput.join('\n');
      // Test might not show ANSI codes due to test environment, so check for colored text or structure
      expect(output).toContain('Colored test');
      expect(output).toContain('🐎');
      expect(output).toContain('[████████████████████░░░░░░░░░░░░░░░░░░░░] 50%');
    });

    test('should work without color manager', () => {
      const plainLoadingScreen = new LoadingScreen(null);
      plainLoadingScreen.renderLoadingScreen('Plain test', '🐎', 0.5);
      
      const output = consoleOutput.join('\n');
      expect(output).toContain('Plain test');
      expect(output).not.toMatch(/\x1b\[[0-9;]*m/); // No ANSI escape sequences
    });
  });

  describe('Transition Display Integration', () => {
    test('should handle displayTransition with proper message sequence', async () => {
      const promise = loadingScreen.displayTransition('main_menu', 'tutorial', 500);
      
      await promise;
      
      const output = consoleOutput.join('\n');
      expect(output).toContain('Loading tutorial...');
    });

    test('should use fallback messages for unknown transitions', async () => {
      const promise = loadingScreen.displayTransition('unknown_state', 'other_state', 500);
      
      await promise;
      
      const output = consoleOutput.join('\n');
      expect(output).toContain('Loading...');
    });
  });
});