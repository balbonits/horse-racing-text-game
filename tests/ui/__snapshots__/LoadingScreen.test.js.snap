// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading Screen System Loading Screen Snapshots should render different animation frames consistently: track-animation-frames 1`] = `
[
  "








                                   üèÅ‚ñë‚ñë‚ñë‚ñë‚ñëüêé

                               Track animation 0

                 [‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë] 0%








",
  "








                                   üèÅ‚ñë‚ñë‚ñëüêé‚ñë‚ñë

                               Track animation 1

                 [‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë] 20%








",
  "








                                   üèÅ‚ñë‚ñëüêé‚ñë‚ñë‚ñë

                               Track animation 2

                 [‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë] 40%








",
  "








                                   üèÅ‚ñëüêé‚ñë‚ñë‚ñë‚ñë

                               Track animation 3

                 [‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë] 60%








",
  "








                                   üèÅüêé‚ñë‚ñë‚ñë‚ñë‚ñë

                               Track animation 4

                 [‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë] 80%








",
]
`;

exports[`Loading Screen System Loading Screen Snapshots should render loading screen consistently: loading-screen-basic 1`] = `
"








                                       üêé

                                Loading test...

                 [‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë] 50%








"
`;

exports[`Loading Screen System Loading Screen Snapshots should render race countdown consistently: race-countdown-sequence 1`] = `
[
  "









",
  "                                       3   ",
  "









",
]
`;

exports[`Loading Screen System Loading Screen Snapshots should render tutorial loading messages consistently: tutorial-loading-messages 1`] = `
[
  "








                                       üêé

                              Starting tutorial...

                 [‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë] 33%








",
  "








                                       üêé

                             Alex Morgan arrives...

                 [‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë‚ñë] 66%








",
  "








                                       üêé

                            Preparing lesson plan...

                [‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà‚ñà] 100%








",
]
`;
